prow/git/v2/client_factory.go

```diff
@@ -391,9 +377,23 @@ func (c *clientFactory) ensureFreshPrimary(cacheDir string, cacheClientCacher ca
 		// This is the default behavior if FetchCommits is empty or nil (i.e.,
 		// when we don't define a targeted list of commits to fetch directly).
 		if repoOpts.FetchCommits.Len() == 0 {
+			defer repoLock.Unlock()
 			if err := cacheClientCacher.RemoteUpdate(); err != nil {
 				return err
 			}
+		} else {
+			// For targeted fetches by SHA objects, there's no need to hold a
+			// lock on the primary because it's safe to do so (git will first
+			// write to a temporary file and replace the file being written to,
+			// so if another git process already wrote to it, the worst case is
+			// that it will overwrite the file with the same data).
+			repoLock.Unlock()
+			// Targeted fetch. Only fetch those commits which we want, and only
+			// if they are missing.
+			if err := cacheClientCacher.FetchCommits(repoOpts.NoFetchTags, repoOpts.FetchCommits.UnsortedList()); err != nil {
+				return err
+			}
+
```

Comment:
The reason I didn't do this is because we still hold the lock for calling RemoteUpdate().

But, I agree that having a single defer is easier to read. I could create the maybeCreatePrimary function and put the call to RemoteUpdate() in there. But then a better name would probably be maybeCloneAndUpdatePrimary. WDYT?

Reply:
Ah, I see that FetchCommits is the alternative to RemoteUpdate(). So you could do like you said and gate the FetchCommits outside the call. I think this would be easier to read and maintain still.

Reply:
Done.